Yoast/wordpress-seo
# Use first image in content as a fallback in the social previews

## Context


* In https://github.com/Yoast/wordpress-seo/issues/14695 some og:image extraction from the content was reinstated. This should be reflected in the social previews.

## Summary


This PR can be summarized in the following changelog entry:

* The new React social previews did not reflect images in the content. This has been reintroduced in wordpress-seo, so we have reintroduced that functionality in the new previews.

## Relevant technical choices:

* We decided that analysing the content in the social preview containers was too resource-heavy: every prop change would trigger a new search for images, either via the tinyMCE API or via the Gutenberg store.
* Instead, we have hooked into existing code that kept track of content images for the snippet preview, and pushed the relevant data to our own redux store.
* Now, we can safely select the fallback image from our redux store.

## Test instructions

This PR can be tested by following these steps:

**NOTE:** the expected hierarchy for the social images is as follows:
1. The image set in the social preview tab on the post/term.
2. For Twitter, if no image is set for Twitter specifically, it should be the image that is set for Facebook.
3. The feature image.
4. The first image in the content.
5. The image set in SEO>social>Facebook>Default settings>Image URL.

* **Test this in premium.**
* Merge this branch from the free remote into your local premium feature branch without commiting (`--no-commit` flag).
* Have the monorepo linked on the social feature branch.
* Set a sitewide og image in Yoastseo > social > facebook > default settings > image url.
* Open a new in the block editor, and observe the social previews. The images should be the image you just set.
* Now, add an image in the content. The images in your previews should update to reflect that image.
* Add another image above the image just added. The images in your previews should update again.
* Set a featured image in the sidebar. This should override the content images in your social previews.
* Now set an image in the Facebook preview. This should override the featured image in BOTH social previews.
* Now set a twitter image, this should override the Twitter image.
* Remove the images you set in the previews and the featured image.
* This should show the content images again.
* Switching, deleting, adding, should all have the expected result.
* Install the classic editor plugin, and repeat the process.
* All expectations should be the same.

## UI changes

## Documentation

## Quality assurance


Fixes https://github.com/Yoast/wordpress-seo/issues/15379
